normalized-cache-incubating/src/commonMain/kotlin/com/apollographql/cache/normalized/internal/DefaultApolloStore.kt

```diff
@@ -118,11 +192,111 @@ internal class DefaultApolloStore(
+        .cacheInfo(
+            CacheInfo.Builder()
+                .fromCache(true)
+                .cacheHit(errors.isEmpty())
+                .stale(batchReaderData.cacheHeaders.headerValue(ApolloCacheHeaders.STALE) == "true")
+                .build()
+        )
+        .build()
+  }
+
+  /**
+   * If a position contains an Error, replace it by a null if the field's type is nullable, bubble up if not.
+   */
+  private fun nullBubble(dataWithErrors: Any?, field: CompiledField, errors: MutableList<Error>): Any? {
+    return when (dataWithErrors) {
+      is Map<*, *> -> {
+        @Suppress("UNCHECKED_CAST")
+        dataWithErrors as Map<String, Any?>
+        dataWithErrors.mapValues { (key, value) ->
+          val selection = field.fieldSelections()[key]
+              ?: // Scalar
+              return@mapValues value
+          when (value) {
+            is Error -> {
+              errors.add(value)
+              if (selection.type is CompiledNotNullType) {
+                return null
+              }
+              null
+            }
+
+            else -> {
+              nullBubble(value, selection, errors).also {
+                if (it == null && selection.type is CompiledNotNullType) {
+                  return null
+                }
+              }
+            }
+          }
+        }
+      }
+
+      is List<*> -> {
+        dataWithErrors.mapIndexed { index, value ->
+          val listType = if (field.type is CompiledNotNullType) {
+            (field.type as CompiledNotNullType).ofType as? CompiledListType
+          } else {
+            field.type as? CompiledListType
+          }
+          if (listType == null) {
+            // Scalar
+            return@mapIndexed value
+          }
+          val elementType = listType.ofType
+          when (value) {
+            is Error -> {
+              errors.add(value)
+              if (elementType is CompiledNotNullType) {
+                return null
+              }
+              null
+            }
+
+            else -> {
+              nullBubble(value, field, errors).also {
+                if (it == null && elementType is CompiledNotNullType) {
+                  return null
+                }
+              }
+            }
+          }
+        }
+      }
+
+      else -> {
+        dataWithErrors
+      }
+    }
+  }
+
+  private fun CompiledSelection.fieldSelections(): Map<String, CompiledField> {
+    fun CompiledSelection.fieldSelections(): List<CompiledField> {
+      return when (this) {
+        is CompiledField -> selections.filterIsInstance<CompiledField>() + selections.filterIsInstance<CompiledFragment>()
+            .flatMap { it.fieldSelections() }
+
+        is CompiledFragment -> selections.filterIsInstance<CompiledField>() + selections.filterIsInstance<CompiledFragment>()
+            .flatMap { it.fieldSelections() }
+      }
+    }
+    return fieldSelections().associateBy { it.responseName }
```

Comment:
Feels like this could be losing some merged fields?

```graphql
{
  user {
    name
  }
  # this takes precedence over the previous user?
  user {
    email
  }
}
```

Reply:
Very good catch! 🙏  I tweaked to merge the selections.